Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True

mikaylagawarecki opened the pull request:
## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`. This option relies  on the previous PR in this stack, where storage order was changed to non lexicographical. A `.format_version` entry was added to the zipfile and `calculate_storage_offsets` will only work on checkpoints with `.format_version`.

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.



https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## How does this work

The format for the checkpoint is as such

```
archive_name/
|_ data.pkl
|_.format_version
|_byteorder
|_data/
  |_ 0
  |_ 1
  |_ 2
  |_ ...
|_
```

Each `data/i` record represents a storage, where storages are written in the order that the Pickler encounters them.

For each storage, our `persistent_load` logic saves the following metadata to the pickle file `dtype, numel, key, location` where `numel` is the number of bytes in the storage.

Note that we always use `miniz` writer  in the zip64 mode per [here](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/caffe2/serialize/inline_container.cc#L701) A zipfile record written by miniz looks as such 

```
 ---------------- ----------------- ------------------- ---------------- --------- ------------------------------
| 30 byte header | n byte filename | zip64_extra_data | m byte padding | storage | 16 or 24 byte local dir footer  |
 ---------------- ----------------- ------------------- ---------------- --------- ------------------------------
```

- The header size (30) is given by [`MZ_ZIP_LOCAL_DIR_HEADER_SIZE`](https://github.com/pytorch/pytorch/blob/main/third_party/miniz-3.0.2/miniz.c?fbclid=IwZXh0bgNhZW0CMTEAAR2O8Vysd--UoSCxW70gabXIS1dbz733oHwuUQ5_Ff1hY2WU6PL2i6CSH4A_aem_J9oaU2HpDeWtJKOU9EnVqw#L3290)
- filename will be `"{archive_name}/{filepath}"`

- `zip64_extra_data` is determined by [`mz_zip_writer_create_zip64_extra_data`](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6202). Note that [we only create zip64_extra_data if storage_size >= 0xFFFFFFFF or the offset of the start of the header >= 0xFFFFFFFF](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6519-L6524)
- `m` is determined by [`getPadding`](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/caffe2/serialize/inline_container.cc#L254), which accounts for filename, zip64_extra_data to determine `m` such that the start of `storage` is aligned to 64 bytes. The `m` bytes will always start with `F B padding_size" as the first 4 bytes
- The local dir footer size is determined based on [this snippet ](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6610-L6632): if the buffer size is 0 it is skipped. If the zip64_extra_data was created, it is 24, otherwise it is 16.


When `torch.utils.serialization.config.load.calculate_storage_offsets` is set we do the following
- We keep track of where the "cursor" is in the file using `current_offset`, after each persistent_load call, it will be at the offset where the header for the next record starts
- for the 0th storage, "data/0", we use the regular get_record_offset to determine the start of the storage
- for any other storage, (where the storages will be in order encountered by the unpickler, 0, 1, 2, 3, ...) we use `get_record_offset_no_read`, which re-uses the `getPadding` logic to determine the offset of the storage
- Note that `load_tensor` will only ever be called again with the same key if the storage's `._data_ptr()` is 0 [[pointer1](https://github.com/pytorch/pytorch/blob/main/torch/serialization.py#L1917-L1918)][[pointer2](https://github.com/pytorch/pytorch/blob/main/torch/serialization.py#L1936-L1937)], so we cache the offsets for this edge case
- After each storage, if the storage is non-zero, we account for the local dir footer based on the logic described above 


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143880
* #143879



cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

mikaylagawarecki committed:
Reduce random reads from get_record_offset when loading with mmap=True

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67673026).

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67673026).

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67673026).

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.

https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


Testing strategy:

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.

https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


Testing strategy:

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki commented on test/test_serialization.py:
```diff
@@ -45,6 +45,7 @@
     BytesIOContext,
     download_file,
     instantiate_parametrized_tests,
+    # IS_CI,
```
will uncomment this and the below `skipIf(not IS_CI, ...` before merging

albanD commented on torch/serialization.py:
```diff
@@ -1886,15 +1888,79 @@ def _load(
+
+    def _get_offset(key, name, numel):
+        """
+        Return the offset of the storage associated with key with record name `name` and size numel.
```
nit: add a warning here and in C++ mentioning that the two must be kept in sync !

albanD commented on torch/serialization.py:
```diff
@@ -1886,15 +1888,79 @@ def _load(
             storage = torch.UntypedStorage(nbytes, device="meta")
         elif overall_storage is not None:
-            storage_offset = zip_file.get_record_offset(name)
+            if has_version and calculate_storage_offsets:
```
Check version number? Or since this is valid for any version from now on, checking existence is enough?

albanD commented on torch/serialization.py:
```diff
@@ -1886,15 +1888,79 @@ def _load(
+            if has_version and calculate_storage_offsets:
+                storage_offset = _get_offset(key, name, numel)
+                if run_debug_asserts:
+                    assert (
```
Raise nice RuntimeError here.
Also mention in the error message that this is a debug assert and the env variable that triggered it.

albanD reviewed (commented):
Sounds good!
Only a few nits

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.

https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.

https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.

https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

huydhn commented:
@pytorchbot revert 'Sorry for reverting your change, but either this PR or the base PR breaks distributed tests' -c nosignal

distributed/fsdp/test_fsdp_dtensor_state_dict.py::TestFSDPWithDeviceMeshAndDTensorCUDA::test_dtensor_sharded_optim_load_state_dict_offload_to_cpu_False_is_even_sharded_model_False_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/13000684319/job/36260263155) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/db3685a35cdce32622ab89f6c92e09d52210ff53)

(Note to myself, this is probably a target determination miss)

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but either this PR or the base PR breaks distributed tests' -c nosignal

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`. This option relies  on the previous PR in this stack, where storage order was changed to non lexicographical. A `.format_version` entry was added to the zipfile and `calculate_storage_offsets` will only work on checkpoints with `.format_version`.

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.



https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`. This option relies  on the previous PR in this stack, where storage order was changed to non lexicographical. A `.format_version` entry was added to the zipfile and `calculate_storage_offsets` will only work on checkpoints with `.format_version`.

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.



https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add option to serialization config to reduce random reads from get_record_offset when loading with mmap=True"

## Background

This PR adds `torch.utils.serialization.config.load.calculate_storage_offsets`. This option relies  on the previous PR in this stack, where storage order was changed to non lexicographical. A `.format_version` entry was added to the zipfile and `calculate_storage_offsets` will only work on checkpoints with `.format_version`.

When this is turned on, for `torch.load(mmap=True)`, offsets of each storage record (other than the 0th storage will be calculated instead of relying on `miniz` APIs to determine this).

The existing APIs will issue multiple random reads (reading the end of central directory record, then reading the zipfile header for the record) to determine the storage offset where the record starts. This can greatly degrade `torch.load(mmap=True)` performance for non-filesystem cases.



https://github.com/pytorch/pytorch/blob/6aaae9d78f0992ac6265552e4f8323ef11d62bb0/caffe2/serialize/inline_container.cc#L589-L605


## How does this work

The format for the checkpoint is as such

```
archive_name/
|_ data.pkl
|_.format_version
|_byteorder
|_data/
  |_ 0
  |_ 1
  |_ 2
  |_ ...
|_
```

Each `data/i` record represents a storage, where storages are written in the order that the Pickler encounters them.

For each storage, our `persistent_load` logic saves the following metadata to the pickle file `dtype, numel, key, location` where `numel` is the number of bytes in the storage.

Note that we always use `miniz` writer  in the zip64 mode per [here](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/caffe2/serialize/inline_container.cc#L701) A zipfile record written by miniz looks as such 

```
 ---------------- ----------------- ------------------- ---------------- --------- ------------------------------
| 30 byte header | n byte filename | zip64_extra_data | m byte padding | storage | 16 or 24 byte local dir footer  |
 ---------------- ----------------- ------------------- ---------------- --------- ------------------------------
```

- The header size (30) is given by [`MZ_ZIP_LOCAL_DIR_HEADER_SIZE`](https://github.com/pytorch/pytorch/blob/main/third_party/miniz-3.0.2/miniz.c?fbclid=IwZXh0bgNhZW0CMTEAAR2O8Vysd--UoSCxW70gabXIS1dbz733oHwuUQ5_Ff1hY2WU6PL2i6CSH4A_aem_J9oaU2HpDeWtJKOU9EnVqw#L3290)
- filename will be `"{archive_name}/{filepath}"`

- `zip64_extra_data` is determined by [`mz_zip_writer_create_zip64_extra_data`](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6202). Note that [we only create zip64_extra_data if storage_size >= 0xFFFFFFFF or the offset of the start of the header >= 0xFFFFFFFF](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6519-L6524)
- `m` is determined by [`getPadding`](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/caffe2/serialize/inline_container.cc#L254), which accounts for filename, zip64_extra_data to determine `m` such that the start of `storage` is aligned to 64 bytes. The `m` bytes will always start with `F B padding_size" as the first 4 bytes
- The local dir footer size is determined based on [this snippet ](https://github.com/pytorch/pytorch/blob/7796e308d0636bcbfb2490c80291edd440d4bc42/third_party/miniz-3.0.2/miniz.c#L6610-L6632): if the buffer size is 0 it is skipped. If the zip64_extra_data is created, it is 24, otherwise it is 16.


When `torch.utils.serialization.config.load.calculate_storage_offsets` is set we do the following
- We keep track of where the "cursor" is in the file using `current_offset`, after each persistent_load call, it will be at the offset where the header for the next record starts
- for the 0th storage, "data/0", we use the regular get_record_offset to determine the start of the storage
- for any other storage, (where the storages will be in order encountered by the unpickler, 0, 1, 2, 3, ...) we use `get_record_offset_no_read`, which re-uses the `getPadding` logic to determine the offset of the storage
- Note that `load_tensor` will only ever be called again with the same key if the storage's `._data_ptr()` is 0 [[pointer1](https://github.com/pytorch/pytorch/blob/main/torch/serialization.py#L1917-L1918)][[pointer2](https://github.com/pytorch/pytorch/blob/main/torch/serialization.py#L1936-L1937)], so we cache the offsets for this edge case
- After each storage, if the storage is non-zero, we account for the local dir footer based on the logic described above 


## Testing strategy

The agreed upon testing strategy was as follows:
- Add debug code gated by an environment flag `TORCH_SERIALIZATION_DEBUG` that will run this offset calculation logic and verify it against getRecordOffset for each storage (when mmap=False)
- This flag is set throughout CI, which means that every time `torch.load` is called, the offset calculation logic is implicitly being tested.




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D67673026](https://our.internmc.facebook.com/intern/diff/D67673026)

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the fix!

mikaylagawarecki commented:
@pytorchbot merge
